### ISSUE TITLE
Various class based NamedTuple and TypedDict errors not caught

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#6200]
This example has a bunch of errors that the semantic analyzer should probably report (and one what seems like a false positive):

```py
from typing import NamedTuple, TypeVar, Generic
from typing_extensions import Protocol
from mypy_extensions import TypedDict

@xyz  # No error
class N1(NamedTuple):
   x: int

@xyz  # No error
class T1(TypedDict):
    x: int

class N2(NamedTuple, Protocol):  # No error
   x: int

class T2(TypedDict, Protocol):  # No error
    x: int

class M(type):
    def f(cls) -> None: pass

class N3(NamedTuple, metaclass=M):  # No error
   x: int

N3.f()  # Error

class T3(TypedDict, metaclass=M):  # No error
    x: int

T = TypeVar('T')

class N4(NamedTuple, Generic[T]):  # No error
   x: int

class T4(TypedDict, Generic[T]):  # No error
   x: int
```

**Comment by AlexWaygood** [comment]
This one now (correctly) causes mypy to emit an error:

```python 
@xyz  # error: Name "xyz" is not defined
class T1(TypedDict):
    x: int
```

The others all [continue to reproduce](https://mypy-play.net/?mypy=latest&python=3.10&flags=warn-unreachable&gist=4e5bd485912a7b2f570c2740adbcc304) on 0.941.
